README.md
# color_extract
[![pub package](https://img.shields.io/pub/v/color_extract.svg)](https://pub.dartlang.org/packages/color_extract) [![Build](https://github.com/itisnajim/color_extract/workflows/Main/badge.svg)](https://github.com/itisnajim/color_extract/actions) [![codecov](https://codecov.io/gh/itisnajim/color_extract/branch/main/graph/badge.svg?token=DQTMJA22JQ)](https://codecov.io/gh/itisnajim/color_extract) [![GitHub license](https://img.shields.io/github/license/itisnajim/color_extract)](https://github.com/itisnajim/color_extract/blob/master/LICENSE)  [![GitHub issues](https://img.shields.io/github/issues/itisnajim/color_extract)](https://github.com/itisnajim/color_extract/issues)

Color Extract is a Flutter package that allows you to extract and calculate colors from your app's widgets.
You can use it to change the color of a widget depending/based on the color of the background (the other widget behind).


Preview
------------
(Video player may not show on pub.dev, check github.com)

https://user-images.githubusercontent.com/44414626/220677525-30942250-ceb2-4e85-a8af-1cb2317e4ed5.mp4

Demo
------------
In this demo see how you can make your widget change colors like a chameleon 🦎.

![demo](https://raw.githubusercontent.com/itisnajim/color_extract/main/readme/example-demo.gif)

Installation
------------

Add the following to your `pubspec.yaml`:


```yaml
dependencies:
  color_extract: ^1.0.1
```

Then run `flutter pub get`.

Usage
-----

### ColorExtractor

The `ColorExtractor` it's the widget we want to extract the average color from. It serves as a wrapper for RepaintBoundary, so you can utilize RepaintBoundary as an alternative.

```dart
ColorExtractor(
  boundaryKey: GlobalKey(),
  child: Container(
    width: 200,
    height: 200,
    color: Colors.red,
  ),
);
```

### ColorAverager

The `ColorAverager` widget computes the average color of a specific portion of either `ColorExtractor` or `RepaintBoundary`. Its application is useful in determining the dominant color of a certain area, such as the background behind a logo, image.

```dart
ColorAverager(
  boundaryKey: GlobalKey(),
  child: SizedBox(
    width: 50,
    height: 50,
  ),
  onChanged: (color) {
    // Handle the new average color.
  },
);
```

You can also use the `ColorAveragerController` to calculate the average color programmatically.

```dart
final controller = ColorAveragerController();

// ... render the widget ...

final avgColor = await controller.calculateAvgColor();
```

Example
-------

```dart
import 'package:flutter/material.dart';
import 'package:color_extract/color_extract.dart';

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          ColorExtractor(
            boundaryKey: boundaryKey,
            child: Container(
              width: 200,
              height: 200,
              color: Colors.blue,
            ),
          ),
          ColorAverager(
            // boundaryKey should be the same one in the above ColorExtractor boundaryKey
            boundaryKey: boundaryKey,
            // You can use the controller (ColorAveragerController) too.
            // controller: controller,
            child: const SizedBox(width: 50, height: 50),
            onChanged: (color) {
                // Do something with the average color.
                // color should be = Colors.blue
            },
          )
        ],
      )
    );
  }
}
```

## Author

itisnajim, itisnajim@gmail.com

## License

color_extract is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
## 1.0.1+1

* Added unit tests to improve code quality and ensure functionality.
* Updated README file to provide clearer instructions and information about the package.
* Added documentation (comments) throughout the code to make it easier for others to understand and use.


## 1.0.0

* Initial release of the color_extract package.
* Added ColorExtractor widget for extracting the dominant color from.
* Added ColorAverager widget for averaging the color of an area of the screen.
* Added ColorAveragerController for controlling the behavior of ColorAverager.
* Added example code to demonstrate the usage of the widgets and controller.
* Added unit tests for both widgets and the controller.


test/color_extract_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:color_extract/color_extract.dart';

void main() {
  testWidgets('ColorAverager should calculate the correct average color',
      (WidgetTester tester) async {
    await tester.runAsync(() async {
      // Create a container with a RepaintBoundary widget
      //that has a red background color.
      final colorToTest = getRandomColor();
      final boundaryKey = GlobalKey();
      final controller = ColorAveragerController();
      final content = Stack(
        children: [
          ColorExtractor(
            boundaryKey: boundaryKey,
            child: Container(
              width: 200,
              height: 200,
              color: colorToTest,
            ),
          ),
          ColorAverager(
            boundaryKey: boundaryKey,
            controller: controller,
            child: const SizedBox(width: 50, height: 50),
            onChanged: (_) {},
          )
        ],
      );

      // Render the container and pump the frame to ensure
      //that the RepaintBoundary is painted.
      await tester.pumpWidget(MaterialApp(home: content));
      await tester.pumpAndSettle(const Duration(seconds: 1));
      final avgColor = await controller.calculateAvgColor(false);
      await tester.pumpAndSettle();

      // Ensure that the average color is calculated correctly.
      debugPrint(colorToTest.toString());
      expect(avgColor, colorToTest);
    });
  });

  testWidgets(
      'GrayscaleFilter should convert the blue child widget '
      'to grayscale of the color 0xff848484', (WidgetTester tester) async {
    await tester.runAsync(() async {
      const colorToTest = Color(0xFF2196F3);
      final childWidget = GrayscaleFilter(
        child: Container(
          color: colorToTest,
          width: 100,
          height: 100,
        ),
      );

      final boundaryKey = GlobalKey();
      final controller = ColorAveragerController();

      final content = Stack(
        children: [
          ColorExtractor(
            boundaryKey: boundaryKey,
            child: childWidget,
          ),
          ColorAverager(
            boundaryKey: boundaryKey,
            controller: controller,
            child: const SizedBox(width: 50, height: 50),
            onChanged: (_) {},
          )
        ],
      );

      await tester.pumpWidget(MaterialApp(home: content));

      final avgColor = await controller.calculateAvgColor(false);

      expect(avgColor, const Color(0xff848484));
    });
  });

  test('reverseColor returns the expected value', () {
    const originalColor = Color(0xFF2196F3);
    final reversedColor = reverseColor(originalColor);
    expect(reversedColor, const Color(0xFFDE690C));
    expect(reversedColor, originalColor.reverse);
  });

  test('withLightness method returns correct color', () {
    const originalColor = Colors.blue;
    final newColor = originalColor.withLightness(0.5);
    final expectedColor =
        HSLColor.fromColor(originalColor).withLightness(0.5).toColor();

    expect(newColor, expectedColor);
  });

  test('moreVisibleColor returns a more visible color', () {
    final originalColor =
        const Color(0xFF2196F3).withLightness(0.8); // light gray
    const backgroundColor = Color(0xFF000000); // black
    final newColor = moreVisibleColor(originalColor, backgroundColor);
    const expectedColor = Color(0xFF9ED1FA);

    expect(newColor, isNot(const Color(0xFF2196F3)));
    expect(newColor, expectedColor);
  });

  test('enforceColor returns a color with adjusted brightness', () {
    const originalColor = Color(0xFF1A237E);
    // Because the color is darker, enforceColor will make it even more darker!
    final enforcedColor = enforceColor(originalColor, 0.2);

    expect(enforcedColor, isNot(equals(originalColor)));
    expect(enforcedColor.computeLuminance(),
        lessThan(originalColor.computeLuminance()));
  });
}


lib/src/utils.dart
import 'dart:math';
import 'package:flutter/material.dart';

/// Converts an ABGR color format to ARGB by swapping R and B values.
int abgrToArgb(int argbColor) {
  int r = (argbColor >> 16) & 0xFF;
  int b = argbColor & 0xFF;
  return (argbColor & 0xFF00FF00) | (b << 16) | r;
}

/// A widget that applies a grayscale filter to its child.
class GrayscaleFilter extends StatelessWidget {
  final Widget child;

  const GrayscaleFilter({
    super.key,
    required this.child,
  });

  @override
  Widget build(BuildContext context) {
    return ColorFiltered(
      colorFilter: const ColorFilter.matrix(<double>[
        0.2126,
        0.7152,
        0.0722,
        0,
        0,
        0.2126,
        0.7152,
        0.0722,
        0,
        0,
        0.2126,
        0.7152,
        0.0722,
        0,
        0,
        0,
        0,
        0,
        1,
        0,
      ]),
      child: child,
    );
  }
}

/// Inverts the RGB values of the input color and sets the alpha value to
/// 255 (fully opaque) to produce the resulting color.
/// The input color should have an alpha value of 255.
Color reverseColor(Color color) {
  return Color((0xFFFFFFFF - color.value) | 0xFF000000);
}

/// An extension on the [Color] class that provides additional functionality.
extension ColorExt on Color {
  /// Returns a new color with the same hue and saturation as this color,
  /// but with the specified [lightness] value.
  Color withLightness(double lightness) {
    final hslColor = HSLColor.fromColor(this);
    return hslColor.withLightness(lightness).toColor();
  }

  // Returns a new color by reversing the color value
  // of the current [Color] instance.
  Color get reverse => reverseColor(this);
}

/// Returns a color that is more visible against the specified background color.
///
/// If the brightness difference between the foreground and background colors
/// is greater than [percent], the foreground color is returned. Otherwise,
/// a new color is returned with a brightness value that is more
/// visible against the background color.
Color moreVisibleColor(
  Color foregroundColor,
  Color backgroundColor, [
  // percent is the desired brightness diff.
  double percent = 0.4,
]) {
  final foregroundLuminance = foregroundColor.computeLuminance();
  final backgroundLuminance = backgroundColor.computeLuminance();
  final brightnessDiff = foregroundLuminance - backgroundLuminance;

  if (brightnessDiff.abs() >= percent) return foregroundColor;

  final sign = brightnessDiff.isNegative ? -1 : 1;
  final newBrightness = min(max(0, foregroundLuminance + sign * percent), 1);
  return foregroundColor.withLightness(newBrightness.toDouble());
}

/// Returns a color that is closer to the specified color, with a brightness
/// value that is adjusted by [percent].
Color enforceColor(Color color, [double percent = 0.2]) {
  final brightness = color.computeLuminance();
  const threshold = 0.5;

  if (brightness == threshold) return color;

  final newBrightness =
      (brightness + (brightness > threshold ? percent : -percent))
          .clamp(0.0, 1.0);
  return color.withLightness(newBrightness);
}

/// Returns a random color generated using the dart:math library. The color
/// has an alpha value of 255 and random red, green, and blue values.
Color getRandomColor() {
  final random = Random();
  return Color.fromARGB(
    255,
    random.nextInt(256),
    random.nextInt(256),
    random.nextInt(256),
  );
}


lib/src/color_averager.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'utils.dart';

/// This widget calculates the average color of the overlapping area between
/// its child and a specified ColorExtractor or RepaintBoundary widget,
/// identified by its GlobalKey boundaryKey.
class ColorAverager extends StatefulWidget {
  /// Creates a new ColorAverager widget.
  const ColorAverager({
    Key? key,
    required this.boundaryKey,
    required this.child,
    this.onChanged,
    this.controller,
    this.fillerColor,
  })  : assert(onChanged != null || controller != null),
        super(key: key);

  /// The `GlobalKey` of the `RepaintBoundary` or `ColorExtractor` widget
  /// to calculate the average color from.
  final GlobalKey boundaryKey;

  /// The child widget whose intersection with
  /// the `RepaintBoundary` or `ColorExtractor` widget will be used
  /// to calculate the average color.
  final Widget child;

  /// The callback function that returns the calculated average color.
  final ValueChanged<Color?>? onChanged;

  /// An optional `ColorAveragerController` object that can be used
  /// to programmatically calculate the average color.
  final ColorAveragerController? controller;

  /// An optional color to use for pixels with an alpha value of 0.
  final Color? fillerColor;

  @override
  State<ColorAverager> createState() => _ColorAveragerState();
}

class _ColorAveragerState extends State<ColorAverager> {
  Color? _avgColor;
  Color? _lastAvgColor;

  @override
  void initState() {
    super.initState();
    widget.controller?._addState(this);

    // Calculate the initial average color.
    WidgetsBinding.instance.addPostFrameCallback((_) {
      Future.delayed(const Duration(milliseconds: 500), _calculateAvgColor);
    });
  }

  @override
  void didUpdateWidget(ColorAverager oldWidget) {
    super.didUpdateWidget(oldWidget);
    // Recalculate the average color if the boundary key has changed.
    if (oldWidget.boundaryKey != widget.boundaryKey) {
      _calculateAvgColor();
    }
  }

  /// Calculates the average color from the intersection between
  /// the child widget and the `RepaintBoundary` widget.
  ///
  /// If `notify` is `true`, this function will also call the `onChanged`
  /// callback function and update the widget state.
  Future<Color?> _calculateAvgColor([bool notify = true]) async {
    final backgroundBoundary = widget.boundaryKey.currentContext
        ?.findRenderObject() as RenderRepaintBoundary?;
    if (backgroundBoundary == null) return null;

    final childBox = context.findRenderObject() as RenderBox;
    final childOffset = childBox.localToGlobal(Offset.zero);

    final childBoundary = childOffset & childBox.size;
    final backgroundBoundaryBox =
        backgroundBoundary.localToGlobal(Offset.zero) & backgroundBoundary.size;
    final intersection = childBoundary.intersect(backgroundBoundaryBox);

    // If there is no intersection between the child
    //and the background boundary, return null.
    if (intersection.width <= 0 || intersection.height <= 0) return null;

    // Get the image bytes from the background boundary.
    final image = await backgroundBoundary.toImage();
    final bytes = await image.toByteData();
    if (bytes == null) return null;

    // Get the pixel values for the intersection of the child
    //and the background boundary.
    final pixels = bytes.buffer.asUint32List();
    int red = 0, green = 0, blue = 0, alpha = 0, count = 0;
    for (var y = intersection.top.toInt();
        y < intersection.bottom.toInt();
        y++) {
      final pixelRowOffset = y * image.width;
      for (var x = intersection.left.toInt();
          x < intersection.right.toInt();
          x++) {
        final pixelOffset = pixelRowOffset + x;

        final pixel = pixels[pixelOffset];
        final argbColor = abgrToArgb(pixel);
        final pixelAlpha = pixel >> 24 & 0xFF;
        final color = pixelAlpha == 0 && widget.fillerColor != null
            ? widget.fillerColor!
            : Color(argbColor);
        red += color.red;
        green += color.green;
        blue += color.blue;
        alpha += color.alpha;
        count++;
      }
    }

    // Calculate the average color from the pixel values.
    if (count == 0) return null;

    final avgRed = red ~/ count;
    final avgGreen = green ~/ count;
    final avgBlue = blue ~/ count;
    final avgAlpha = alpha ~/ count;

    final avgColor = Color.fromRGBO(avgRed, avgGreen, avgBlue, avgAlpha / 255);
    if (_lastAvgColor == avgColor) return avgColor;

    if (notify) {
      // Update the state with the new average color and notify the callback.
      setState(() {
        _avgColor = avgColor;
        _lastAvgColor = avgColor;
        widget.onChanged?.call(_avgColor);
      });
    }

    return avgColor;
  }

  @override
  Widget build(BuildContext context) {
    return widget.child;
  }
}

/// Controller class for [ColorAverager].
class ColorAveragerController {
  _ColorAveragerState? _widgetState;
  void _addState(_ColorAveragerState widgetState) {
    _widgetState = widgetState;
  }

  /// Calculates the average color and updates the [ColorAverager] object state.
  ///
  /// If notify is true, it notifies the [onChanged] callback function
  /// with the calculated average color.
  Future<Color?> calculateAvgColor([bool notify = true]) {
    return _widgetState!._calculateAvgColor(notify);
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:color_extract/color_extract.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/scheduler.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class PosColor {
  Offset offset;
  Size size;
  Color color;

  PosColor({
    required this.offset,
    required this.size,
    required this.color,
  });
}

List<PosColor> generatePosColors() {
  List<PosColor> posColors = [];
  final random = Random();

  for (int i = 0; i < 20; i++) {
    double width = random.nextDouble() * 200 + 100;
    double height = random.nextDouble() * 200 + 100;
    double x = random.nextDouble() * 300 + 50;
    double y = random.nextDouble() * 800 + 50;

    posColors.add(
      PosColor(
        offset: Offset(x, y),
        size: Size(width, height),
        color: getRandomColor(),
      ),
    );
  }

  for (int i = 0; i < 5; i++) {
    double size = random.nextDouble() * 100 + 50;
    double x = 250 - size / 2;
    double y = 400 - size / 2;

    posColors.add(
      PosColor(
        offset: Offset(x, y),
        size: Size(size, size),
        color: Colors.white,
      ),
    );
  }

  for (int i = 0; i < 5; i++) {
    double size = random.nextDouble() * 100 + 50;
    double x = 250 - size / 2;
    double y = 550 - size / 2;

    posColors.add(
      PosColor(
        offset: Offset(x, y),
        size: Size(size, size),
        color: Colors.black,
      ),
    );
  }
  for (int i = 0; i < 20; i++) {
    double width = random.nextDouble() * 200 + 100;
    double height = random.nextDouble() * 200 + 100;
    double x = random.nextDouble() * 300 + 50;
    double y = random.nextDouble() * 800 + 600;

    posColors.add(
      PosColor(
        offset: Offset(x, y),
        size: Size(width, height),
        color: getRandomColor(),
      ),
    );
  }

  return posColors;
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final boundaryKey = GlobalKey();
  final _scrollController = ScrollController();
  final _controller = ColorAveragerController();

  final recsInfo = ValueNotifier(generatePosColors());
  final draggableInfo = ValueNotifier(
    PosColor(
        offset: const Offset(30, 30),
        color: Colors.blue,
        size: const Size(50, 50)),
  );

  Future<Color> calcAvgColor([bool reverse = false]) async {
    var color = (await _controller.calculateAvgColor()) ?? Colors.blue;
    if (reverse) color = color.reverse;
    return color;
  }

  bool isReversedColor = false;

  @override
  void initState() {
    _scrollController.addListener(() {
      if (_scrollController.hasClients) {
        if (_scrollController.position.userScrollDirection !=
            ScrollDirection.idle) {
          SchedulerBinding.instance.addPostFrameCallback((_) {
            _controller.calculateAvgColor();
          });
        }
      }
    });

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          ColorExtractor(
            boundaryKey: boundaryKey,
            child: SingleChildScrollView(
              controller: _scrollController,
              child: SizedBox(
                height: 2000,
                child: Stack(
                    children: recsInfo.value
                        .map((posColor) => Positioned(
                              left: posColor.offset.dx,
                              top: posColor.offset.dy,
                              child: Container(
                                width: posColor.size.width,
                                height: posColor.size.height,
                                color: posColor.color,
                              ),
                            ))
                        .toList()),
              ),
            ),
          ),
          ValueListenableBuilder(
            valueListenable: draggableInfo,
            builder: (_, __, ___) {
              return Positioned(
                left: draggableInfo.value.offset.dx,
                top: draggableInfo.value.offset.dy,
                child: Draggable(
                  feedback: Opacity(
                    opacity: 0.6,
                    child: Container(
                      width: draggableInfo.value.size.width,
                      height: draggableInfo.value.size.height,
                      decoration: BoxDecoration(
                        color: draggableInfo.value.color,
                        borderRadius: BorderRadius.circular(8),
                        border: Border.all(color: Colors.grey.withOpacity(0.4)),
                      ),
                    ),
                  ),
                  onDragEnd: (dragDetails) async {
                    draggableInfo.value = PosColor(
                      color: await calcAvgColor(isReversedColor),
                      offset: Offset(
                        dragDetails.offset.dx,
                        dragDetails.offset.dy -
                            MediaQuery.of(context).padding.top,
                      ),
                      size: draggableInfo.value.size,
                    );
                    WidgetsBinding.instance.addPostFrameCallback((_) async {
                      draggableInfo.value = PosColor(
                        color: await calcAvgColor(isReversedColor),
                        offset: Offset(
                          dragDetails.offset.dx,
                          dragDetails.offset.dy -
                              MediaQuery.of(context).padding.top,
                        ),
                        size: draggableInfo.value.size,
                      );
                    });
                  },
                  child: ColorAverager(
                    boundaryKey: boundaryKey,
                    controller: _controller,
                    fillerColor: Theme.of(context).scaffoldBackgroundColor,
                    child: AnimatedContainer(
                      duration: const Duration(milliseconds: 300),
                      width: draggableInfo.value.size.width,
                      height: draggableInfo.value.size.height,
                      decoration: BoxDecoration(
                        color: draggableInfo.value.color,
                        borderRadius: BorderRadius.circular(8),
                        border: Border.all(color: Colors.grey.withOpacity(0.4)),
                      ),
                    ),
                    onChanged: (value) {
                      //debugPrint('value $value');
                      final color = value ?? Colors.blue;
                      draggableInfo.value = PosColor(
                        color: isReversedColor ? color.reverse : color,
                        offset: draggableInfo.value.offset,
                        size: draggableInfo.value.size,
                      );
                    },
                  ),
                ),
              );
            },
          ),
          Positioned(
              bottom: 0,
              left: 0,
              right: 0,
              child: ColoredBox(
                color:
                    Theme.of(context).scaffoldBackgroundColor.withOpacity(0.8),
                child: CheckboxListTile(
                  controlAffinity: ListTileControlAffinity.trailing,
                  title: const Text('Reverse Color'),
                  value: isReversedColor,
                  onChanged: (_) {
                    setState(() {
                      isReversedColor = !isReversedColor;
                      WidgetsBinding.instance.addPostFrameCallback((_) async {
                        draggableInfo.value = PosColor(
                          color: await calcAvgColor(isReversedColor),
                          offset: draggableInfo.value.offset,
                          size: draggableInfo.value.size,
                        );
                      });
                    });
                  },
                ),
              ))
        ],
      ),
    );
  }
}
